### PR TITLE
Add exponential reconnection backoff, solves #140

### PR DESCRIPTION
**Pull request by Frotty**
See https://github.com/kiwiirc/irc-framework/issues/140

### CODE REVIEW

**Comment by prawnsalad** [comment]
That jitter value should be much larger.. probably in the range of 1-5 seconds even. Then it should also have a max timeout length so the timeout doesn't get insanely huge. Something like max of 5 minutes sounds reasonable IMO.

File: src/connection.js [hunk 1]
```diff
@@ -136,30 +136,33 @@ module.exports = class Connection extends EventEmitter {
             }
 
             if (should_reconnect) {
+                var reconnect_wait = that.calculateExponentialBackoff();
```
Comment (xPaw) [inline src/connection.js]: Use let/const please (outside of the if)

File: src/connection.js [hunk 2]
```diff
@@ -136,30 +136,33 @@ module.exports = class Connection extends EventEmitter {
...
                 that.reconnect_attempts++;
                 that.emit('reconnecting', {
                     attempt: that.reconnect_attempts,
                     max_retries: that.auto_reconnect_max_retries,
-                    wait: that.auto_reconnect_wait
+                    wait: reconnect_wait
                 });
+                
+                that.debugOut('Scheduling reconnect');
+                that.setTimeout(function() {
```
Comment (xPaw) [inline src/connection.js]: Since you moved it anyway: 
```suggestion
                that.setTimeout(() => that.connect, reconnect_wait);
```

**Comment by Frotty** [comment]
@prawnsalad Agreed on the jitter. For the maximum I would prefer a higher one, since my actual use case was getting g-lined from quakenet, which takes about an hour to clear up.

File: src/connection.js [hunk 3]
```diff
@@ -136,26 +136,34 @@ module.exports = class Connection extends EventEmitter {
...
+                
                 that.reconnect_attempts++;
                 that.emit('reconnecting', {
                     attempt: that.reconnect_attempts,
                     max_retries: that.auto_reconnect_max_retries,
-                    wait: that.auto_reconnect_wait
+                    wait: reconnect_wait
                 });
+                
+                that.debugOut('Scheduling reconnect');
```
Comment (xPaw) [inline src/connection.js]: Would be reasonable to put the wait time in here for debugging, and maybe the attemps/max tries too.

**Comment by xPaw** [comment]
Looks good, will need to give it some testing.

**Comment by Frotty** [comment]
@prawnsalad Anything left to do here?

**Comment by prawnsalad** [comment]
After trying this out the reconnect timeouts (`reconnect_attempts++` between each call) gives:
~~~
13657
14954
19766
37475
68025
135704
260587
~~~

That's a bit high and clustered at the beginning so I tweaked it to:
~~~javascript
function calculateExponentialBackoff() {
    const jitter = 1000 + Math.floor(Math.random() * 5000);
    const attempts = Math.min(reconnect_attempts, 30);
    const time = 1000 * Math.pow(2, attempts);
    return Math.min(time, auto_reconnect_max_wait) + jitter;
}
~~~

Which gives a better result:
~~~
3077
7235
8395
11505
21760
36047
65086
~~~

This has also removed the redundant `auto_reconnect_wait` option.

**Comment by xPaw** [comment]
@Frotty are you able to rebase/squash this and apply the tweak above?

File: src/connection.js [hunk 4]
```diff
@@ -158,10 +157,10 @@ module.exports = class Connection extends EventEmitter {
...
     
     calculateExponentialBackoff() {
-        const jitter = 1000 + Math.floor(Math.random() * 9000);
-        const exponent = Math.min(this.reconnect_attempts, 30);
-        const time = this.auto_reconnect_wait * Math.pow(2, exponent);
-        return Math.min(time, this.auto_reconnect_max_wait) + jitter;
+        const jitter = 1000 + Math.floor(Math.random() * 5000);
+        const attempts = Math.min(reconnect_attempts, 30);
+        const time = 1000 * Math.pow(2, attempts);
+        return Math.min(time, auto_reconnect_max_wait) + jitter;
```
Comment (xPaw) [inline src/connection.js:163]: I believe the `this.` prefix needs to stay (and above)

**Comment by prawnsalad** [comment]
Nice, will get to testing it tonight/tomorrow ++

**Pull request merged**